plugins/osdn/osdn.go

```diff
@@ -100,32 +104,32 @@ func (oi *OsdnRegistryInterface) CreateSubnet(minion string, sub *osdnapi.Subnet
 }
 
 func (oi *OsdnRegistryInterface) WatchSubnets(receiver chan *osdnapi.SubnetEvent, stop chan bool) error {
-	// double go :(
-	revision := ""
-	wi, err := oi.oClient.HostSubnets().Watch(revision)
-	if err != nil {
-		return err
+	subnetEventQueue := oscache.NewEventQueue(cache.MetaNamespaceKeyFunc)
+	listWatch := &cache.ListWatch{
+		ListFunc: func() (runtime.Object, error) {
+			return oi.oClient.HostSubnets().List()
+		},
+		WatchFunc: func(resourceVersion string) (watch.Interface, error) {
+			return oi.oClient.HostSubnets().Watch(resourceVersion)
+		},
 	}
-	go func() {
-		for {
-			ev := <-wi.ResultChan()
-			switch ev.Type {
-			case watch.Added:
-				// create SubnetEvent
-				hs := ev.Object.(*api.HostSubnet)
-				receiver <- &osdnapi.SubnetEvent{Type: osdnapi.Added, Minion: hs.Host, Sub: osdnapi.Subnet{Minion: hs.HostIP, Sub: hs.Subnet}}
-			case watch.Deleted:
-				hs := ev.Object.(*api.HostSubnet)
-				receiver <- &osdnapi.SubnetEvent{Type: osdnapi.Deleted, Minion: hs.Host, Sub: osdnapi.Subnet{Minion: hs.HostIP, Sub: hs.Subnet}}
-			case watch.Modified:
-				hs := ev.Object.(*api.HostSubnet)
-				receiver <- &osdnapi.SubnetEvent{Type: osdnapi.Added, Minion: hs.Host, Sub: osdnapi.Subnet{Minion: hs.HostIP, Sub: hs.Subnet}}
-			case watch.Error:
-				fmt.Errorf("Error in watching subnets")
-				return
-			}
+	cache.NewReflector(listWatch, &api.HostSubnet{}, subnetEventQueue, 4*time.Minute).Run()
+
+	for {
+		eventType, obj, err := subnetEventQueue.Pop()
+		if err != nil {
+			return err
 		}
-	}()
+		switch eventType {
+		case watch.Added, watch.Modified:
+			// create SubnetEvent
+			hs := obj.(*api.HostSubnet)
+			receiver <- &osdnapi.SubnetEvent{Type: osdnapi.Added, Minion: hs.Host, Sub: osdnapi.Subnet{Minion: hs.HostIP, Sub: hs.Subnet}}
+		case watch.Deleted:
```

Comment:
Deletes aren't guaranteed to be delivered, add a todo here that there needs to be a delete controller loop to periodically correct for that.  Ben and Abhisheck have work around that.